src/worker/extensions/echo-message/index.js

```diff
@@ -0,0 +1,62 @@
+const IrcMessage = require('irc-framework').Message;
+
+let baseId = 'kiwibnc-'+Date.now();
+let msgId = 0;
+
+module.exports.init = async function init(hooks) {
+    // echo-message support
+    hooks.on('available_caps', (event) => {
+        event.caps.add('echo-message');
+    });
+    hooks.on('wanted_caps', (event) => {
+        event.wantedCaps.add('echo-message');
+    });
+    hooks.on('message_from_client', (event) => {
+        if (!event.client.state.netRegistered) {
+            return;
+        }
+        let upstream = event.client.upstream;
+        if(!upstream) {
+            return;
+        }
+        // If the server doesn't support echo-message we do it ourselves with our own message.
+        if (!upstream.state.caps.has('echo-message')) {
+            let msg = event.message;
+            if(msg.command !== 'PRIVMSG' && msg.command !== 'NOTICE' && msg.command !== 'ACTION') {
+                return;
+            }
+            // Give ID to original message so it is stored correctly
+            msg.tags.msgid = baseId + '-' + msgId++;
+
+            let m = new IrcMessage(msg.command, ...msg.params);
+            m.tags = {...msg.tags};
+            m.nick = upstream.state.nick;
+            m.username = upstream.state.username;
+            m.hostname = upstream.state.host;
+            m.prefix = m.nick + '!' + m.username + '@' + m.hostname;
+
+            upstream.forEachClient((client) => {
+                // Don't echo back to client that sent if it's not expecting it.
+                if(client === event.client && !event.client.state.caps.has('echo-message')) {
```

Comment:
Should this `&&` be `||` here?

Reply:
I think it's correct. In this case upstream doesn't support echo-message so the bnc is emulating it. If the client that sent the message supported echo-message it would expect to see it's message echoed back to it, if the client doesn't then it wouldn't expect to see it, with an `||` the client that sent it would never receive it's own message back even if it supported echo-message and was expecting it.